src/dataEntities/appointment.ts

```diff
@@ -271,9 +283,21 @@ export class Appointment {
      * Validate property values on the appointment
      * @param log Logger to be used in case of failures
      */
-    public async validate(log: Logger = logger) {
+    public async validate(
+        blockCache: ReadOnlyBlockCache<IBlockStub>,
+        pisaContractAddress: string,
+        log: Logger = logger
+    ) {
         if (this.paymentHash.toLowerCase() !== Appointment.FreeHash) throw new PublicDataValidationError("Invalid payment hash."); // prettier-ignore
 
+        // the start block must be close to the current block +/- 5
+        const currentHead = blockCache.head.number;
+        if(this.startBlock < (currentHead - 5)) throw new PublicDataValidationError(`Start block too low. Start block must be within 5 blocks of the current block ${currentHead}.`); // prettier-ignore
+        if(this.startBlock > (currentHead + 5)) throw new PublicDataValidationError(`Start block too high. Start block must be within 5 blocks of the current block ${currentHead}.`); // prettier-ignore
+        if((this.endBlock - this.startBlock) > 60000) throw new PublicDataValidationError(`Appointment duration too great. Maximum duration between start and block is 60000.`); // prettier-ignore
+        if((this.endBlock - this.startBlock) < 100) throw new PublicDataValidationError(`Appointment duration too small. Minimum duration between start and block is 100.`); // prettier-ignore
+        
+
```

Comment:
The point of hiring Pisa is that you can go offline immediately after; but if you use the natural approach of using the current block as `startBlock` and then going offline, an attacker capable of a re-org of a few blocks might cause the trigger to happen before the `startBlock`, and Pisa wouldn't be watching.

Perhaps the `startBlock` should be allowed to be more in the past; for example if we consider _d_  to be the reorg-safe depth, then we could allow up to `currentBlock - d - 5` in the past.

Any specific reason we limit the `startBlock` to not be far in the future? I might have hired another watchtower for the next 1000 blocks, so I want to hire Pisa from block `currentBlock + 1000`. I don't see issues with that (although we probably don't want appointments that _really_ far in the future).

Reply:
yeah, I figured 5 was enough depth for that.

In the equation currentBlock - d - 5, what is the purpose of the 5? Just margin of error, different clocks? For currentBlock - d - m what values would you choose? I suppose d would have to be around 20? and m 1 or 2?

For not fair in the future, the reason is just because we're still early stage, and free. I dont want us to bank up a load of unpaid appointments forever - so I picked some limit.

Reply:
I think 5 is enough against accidental forks, but definitely not against active attackers.

for "startBlock >= currentBlock - d - m" my idea is the the customer sets `startBlock = currentBlock - d` so `m` should be chosen so that the chance of `m` block being mined during the negotiation of the appointment is negligible; so 5 might be good. Although another concern is that the client might be using a node which fell behind (probably quite common) - which could suggest a significantly larger value of `m` in practice, or some other form of mitigation (but this is definitely out of scope for this ticket).

`d` should be the "no fork bigger than this will ever happen" threshold, so perhaps 20 or 30; we actually have the `maximum-reorg-limit` setting for Pisa, although it might make sense to have a smaller limit in the appointment validation (e.g. we can internally handle bigger reorgs than we declare).

We also want `d + m` to be much smaller than the `maxDepth` of the blockCache (and perhaps we should enforce this explicitly).

So my suggestion is `d = 20` and `m = 5` as good values for now.

Reply:
Ive done d=20 m=3